virtio-devices/src/watchdog.rs

```diff
@@ -0,0 +1,429 @@
+// Copyright 2017 The Chromium OS Authors. All rights reserved.
+// Use of this source code is governed by a BSD-style license that can be
+// found in the LICENSE file.
+
+use super::Error as DeviceError;
+use super::{
+    ActivateError, ActivateResult, EpollHelper, EpollHelperError, EpollHelperHandler, Queue,
+    VirtioCommon, VirtioDevice, VirtioDeviceType, EPOLL_HELPER_EVENT_LAST, VIRTIO_F_VERSION_1,
+};
+use crate::seccomp_filters::{get_seccomp_filter, Thread};
+use crate::{VirtioInterrupt, VirtioInterruptType};
+use anyhow::anyhow;
+use seccomp::{SeccompAction, SeccompFilter};
+use std::fs::File;
+use std::io::{self, Read};
+use std::os::unix::io::{AsRawFd, FromRawFd, RawFd};
+use std::result;
+use std::sync::atomic::AtomicBool;
+use std::sync::{Arc, Barrier, Mutex};
+use std::thread;
+use std::time::Instant;
+use vm_memory::{Bytes, GuestAddressSpace, GuestMemoryAtomic, GuestMemoryMmap};
+use vm_migration::{
+    Migratable, MigratableError, Pausable, Snapshot, SnapshotDataSection, Snapshottable,
+    Transportable,
+};
+use vmm_sys_util::eventfd::EventFd;
+
+const QUEUE_SIZE: u16 = 8;
+const QUEUE_SIZES: &[u16] = &[QUEUE_SIZE];
+
+// New descriptors are pending on the virtio queue.
+const QUEUE_AVAIL_EVENT: u16 = EPOLL_HELPER_EVENT_LAST + 1;
+// Timer expired
+const TIMER_EXPIRED_EVENT: u16 = EPOLL_HELPER_EVENT_LAST + 2;
+
+// Number of seconds to check to see if there has been a ping
+// This needs to match what the driver is using.
+const WATCHDOG_TIMER_INTERVAL: i64 = 15;
```

Comment:
I think we could do better and be more flexible since the timer interval should be changeable from guest userspace, no matter what has been set by default in the driver. That's why I think we could introduce a `virtio_watchdog_config`, holding onto the timeout value. This value can be written from the driver so that we would catch a change in the config from the device.
And from the driver, that means we must implement `.set_timeout()` callback to be notified when the timeout is updated from `watchdog_set_timeout()`.

Reply:
I thought about it but that adds quite a lot more complexity. And AFAICT, when I was playing with this, systemd just wants to use 15s.

Reply:
That's fine, but that might be something VIRTIO maintainers will ask for. Anyway, we can still add this later if needed with a follow up PR.

Reply:
I agree this should eventually be added, the driver should be able to fetch the device's default timer interval as opposed to having an implicit assumption about it.
And yes, this can be done as a second step.